File: mobile-ios/MesadaApp/ViewModels/LoginConviteViewModel.swift
import Foundation

@MainActor
final class LoginConviteViewModel: ObservableObject {
    @Published var codigo: String = ""
    @Published private(set) var carregando = false
    @Published private(set) var mensagemDeErro: String?

    private let autenticacaoRepository: AutenticacaoRepository

    init(autenticacaoRepository: AutenticacaoRepository) {
        self.autenticacaoRepository = autenticacaoRepository
    }

    func resgatarConvite(aoSucesso: @escaping () -> Void) {
        let codigoLimpo = codigo.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !codigoLimpo.isEmpty else { return }

        carregando = true
        mensagemDeErro = nil

        Task {
            let resultado = await autenticacaoRepository.resgatarConvite(codigo: codigoLimpo)
            carregando = false
            switch resultado {
            case .success:
                aoSucesso()
            case .failure:
                mensagemDeErro = "Código inválido ou expirado. Peça um novo convite ao responsável."
            }
        }
    }
}

File: mobile-android/app/src/main/java/com/mesadaapp/mobile/ui/login/LoginConviteViewModel.kt
package com.mesadaapp.mobile.ui.login

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.mesadaapp.mobile.data.repository.AutenticacaoRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class LoginConviteEstado(
    val codigo: String = "",
    val carregando: Boolean = false,
    val mensagemDeErro: String? = null,
)

class LoginConviteViewModel(
    private val autenticacaoRepository: AutenticacaoRepository,
) : ViewModel() {

    private val _estado = MutableStateFlow(LoginConviteEstado())
    val estado: StateFlow<LoginConviteEstado> = _estado.asStateFlow()

    fun aoMudarCodigo(novoValor: String) {
        _estado.update { it.copy(codigo = novoValor, mensagemDeErro = null) }
    }

    fun resgatarConvite(aoSucesso: () -> Unit) {
        val codigo = _estado.value.codigo.trim()
        if (codigo.isBlank()) return

        _estado.update { it.copy(carregando = true, mensagemDeErro = null) }
        viewModelScope.launch {
            autenticacaoRepository.resgatarConvite(codigo)
                .onSuccess {
                    _estado.update { it.copy(carregando = false) }
                    aoSucesso()
                }
                .onFailure {
                    _estado.update {
                        it.copy(
                            carregando = false,
                            mensagemDeErro = "Código inválido ou expirado. Peça um novo convite ao responsável.",
                        )
                    }
                }
        }
    }

    companion object {
        fun factory(autenticacaoRepository: AutenticacaoRepository) = object : ViewModelProvider.Factory {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>): T =
                LoginConviteViewModel(autenticacaoRepository) as T
        }
    }
}
